modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3PerProjectClientManager.java

```diff
@@ -0,0 +1,283 @@
+                clientsHoldersToClose.add(removed);
+            }
+        }
+        // Close stale clients asynchronously without blocking the applier thread
+        if (clientsHoldersToClose.isEmpty() == false) {
+            final var currentClientsCloseListener = new SubscribableListener<Void>();
+            final var previousClientsCloseListener = clientsCloseListener;
+            clientsCloseListener = currentClientsCloseListener;
+            if (previousClientsCloseListener != null && previousClientsCloseListener.isDone() == false) {
+                previousClientsCloseListener.addListener(
+                    ActionListener.running(() -> closeClientsAsync(clientsHoldersToClose, currentClientsCloseListener))
+                );
+            } else {
+                closeClientsAsync(clientsHoldersToClose, currentClientsCloseListener);
+            }
+        }
+    }
+
+    private void closeClientsAsync(List<ClientsHolder> clientsHoldersToClose, ActionListener<Void> listener) {
+        executor.execute(() -> {
+            IOUtils.closeWhileHandlingException(clientsHoldersToClose);
+            listener.onResponse(null);
+        });
+    }
+
+    public AmazonS3Reference client(ProjectId projectId, String clientName) {
+        assert projectId != null && ProjectId.DEFAULT.equals(projectId) == false : projectId;
+        final var clientsHolder = projectClientsHolders.get(projectId);
+        if (clientsHolder == null) {
+            throw new IllegalArgumentException("no s3 client is configured for project [" + projectId + "]");
+        }
+        return clientsHolder.client(clientName);
+    }
+
+    /**
+     * Similar to S3Service#releaseCachedClients but only clears the cache for the given project.
+     * All clients for the project are closed and will be recreated on next access, also similar to S3Service#releaseCachedClients
+     */
+    public void releaseProjectClients(ProjectId projectId) {
+        assert projectId != null && ProjectId.DEFAULT.equals(projectId) == false : projectId;
+        final var old = projectClientsHolders.get(projectId);
+        if (old != null) {
+            old.clearCache();
+        }
+    }
+
+    /**
+     * Shutdown the manager by closing all clients holders. This is called when the node is shutting down.
+     * It attempts to wait (1 min) for any async client closing to complete.
+     */
+    public void close() {
+        IOUtils.closeWhileHandlingException(projectClientsHolders.values());
+        final var currentClientsCloseListener = clientsCloseListener;
+        if (currentClientsCloseListener != null && currentClientsCloseListener.isDone() == false) {
+            // Wait for async clients closing to be completed
+            final CountDownLatch latch = new CountDownLatch(1);
+            currentClientsCloseListener.addListener(ActionListener.running(latch::countDown));
+            try {
+                if (latch.await(1, TimeUnit.MINUTES) == false) {
```

Comment:
Are you sure we need to wait here? With `S3Service` today we just decref the clients and move on, allowing any ongoing uses to finish and shut down separately. Is that not ok here?

If so, why 1 minute? This needs to be configurable in case we need to change it. And if it's ok to proceed after a minute why wait at all?

Reply:
The wait here is to ensure `S3Service` does not close until all asyncly closed clients (from the applier threads) have finished closing per [your comment](https://github.com/elastic/elasticsearch/pull/126584#issuecomment-2796017196). But maybe I misunderstood what you were suggesting?

The 1 minute is meant to be a best effort to close things cleanly similar to the wait for IndicesService to close in 1min in `Stateless.java`. It's unlike to happen. If somehow a close call does stall, it's not fatal and should not prevent progress (`decRef` can still trigger `closeInternal` and close both `S3Client` and `SdkHttpClient` inline which might take a bit longer?). I can make it configurable if necessary.